File: test/csharp/video/api/integration/VideosTest.java
package video.api.integration;

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
import video.api.client.ApiVideoClient;
import video.api.client.api.ApiException;
import video.api.client.api.models.*;

import java.io.File;
import java.util.Collections;
import java.util.HashSet;
import java.util.concurrent.atomic.AtomicLong;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("Integration tests of api.videos() methods")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@EnabledIfEnvironmentVariable(named = "INTEGRATION_TESTS_API_TOKEN", matches = ".+")
public class VideosTest {

    ApiVideoClient apiClient;

    public VideosTest() {
        this.apiClient = new ApiVideoClient(System.getenv().get("INTEGRATION_TESTS_API_TOKEN"));
    }

    @Nested
    @DisplayName("upload by chunk")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class UploadByChunk {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos()
                    .create(new VideoCreatePayload().title("[Java-SDK-tests] upload with chunk")._public(false));
        }

        @Test
        public void uploadVideo() throws ApiException {
            File mp4File = new File(this.getClass().getResource("/assets/sample.mp4").getFile());

            long fileSize = mp4File.length();
            int chunkSize = 1024 * 100;

            apiClient.getHttpClient().setUploadChunkSize(chunkSize);

            AtomicLong totalUploadedAtomic = new AtomicLong(0);
            AtomicLong totalBytesAtomic = new AtomicLong(0);
            AtomicLong chunkCountAtomic = new AtomicLong(0);
            HashSet<Integer> seenChunkNums = new HashSet<>();

            apiClient.videos().upload(testVideo.getVideoId(), mp4File,
                    (bytesWritten, totalBytes, chunkCount, chunkNum) -> {
                        totalUploadedAtomic.set(bytesWritten);
                        totalBytesAtomic.set(totalBytes);
                        chunkCountAtomic.set(chunkCount);
                        seenChunkNums.add(chunkNum);
                    });

            assertThat(totalBytesAtomic.get()).isEqualTo(fileSize);
            assertThat(totalUploadedAtomic.get()).isEqualTo(fileSize);
            assertThat(chunkCountAtomic.get())
                    .isEqualTo(new Double(Math.ceil((float) fileSize / chunkSize)).longValue());
            assertThat(seenChunkNums).containsExactly(1, 2, 3, 4, 5, 6);

            System.out.println(testVideo);
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }

    @Nested
    @DisplayName("upload without chunk")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class Upload {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos()
                    .create(new VideoCreatePayload().title("[Java-SDK-tests] upload without chunk"));
        }

        @Test
        public void uploadVideo() throws ApiException {
            File mp4File = new File(this.getClass().getResource("/assets/sample.mp4").getFile());

            long fileSize = mp4File.length();
            long chunkSize = fileSize + 1;

            apiClient.getHttpClient().setUploadChunkSize(chunkSize);

            AtomicLong totalUploadedAtomic = new AtomicLong(0);
            AtomicLong totalBytesAtomic = new AtomicLong(0);
            AtomicLong chunkCountAtomic = new AtomicLong(0);
            HashSet<Integer> seenChunkNums = new HashSet<>();

            apiClient.videos().upload(testVideo.getVideoId(), mp4File,
                    (bytesWritten, totalBytes, chunkCount, chunkNum) -> {
                        totalUploadedAtomic.set(bytesWritten);
                        totalBytesAtomic.set(totalBytes);
                        chunkCountAtomic.set(chunkCount);
                        seenChunkNums.add(chunkNum);
                    });

            assertThat(totalBytesAtomic.get()).isEqualTo(fileSize);
            assertThat(totalUploadedAtomic.get()).isEqualTo(fileSize);
            assertThat(chunkCountAtomic.get()).isEqualTo(1);
            assertThat(seenChunkNums).containsExactly(1);
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }

    @Nested
    @DisplayName("update")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class Update {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos()
                    .create(new VideoCreatePayload().title("[Java-SDK-tests] video updates"));
        }

        @Test
        public void addMetadata() throws ApiException {

            Video updated = apiClient.videos().update(testVideo.getVideoId(),
                    new VideoUpdatePayload().addMetadataItem(new Metadata("firstKey", "firstValue"))
                            .addMetadataItem(new Metadata("secondKey", "secondValue")));

            assertThat(updated.getMetadata()).containsExactlyInAnyOrder(new Metadata("firstKey", "firstValue"),
                    new Metadata("secondKey", "secondValue"));

            Video updated2 = apiClient.videos().update(testVideo.getVideoId(), new VideoUpdatePayload()
                    .metadata(updated.addMetadataItem(new Metadata("thirdKey", "thirdValue")).getMetadata()));

            assertThat(updated2.getMetadata()).containsExactlyInAnyOrder(new Metadata("firstKey", "firstValue"),
                    new Metadata("secondKey", "secondValue"), new Metadata("thirdKey", "thirdValue"));

            Video updated3 = apiClient.videos().update(testVideo.getVideoId(),
                    new VideoUpdatePayload().metadata(Collections.emptyList()));

            assertThat(updated3.getMetadata()).isEmpty();
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }

    @Nested
    @DisplayName("get")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class Get {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos().create(new VideoCreatePayload().title("[Java-SDK-tests] get"));
        }

        @Test
        public void get() throws ApiException {
            Video video = apiClient.videos().get(testVideo.getVideoId());

            assertThat(video.getVideoId()).isEqualTo(testVideo.getVideoId());
            assertThat(video.getTitle()).isEqualTo(testVideo.getTitle());
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }

    @Nested
    @DisplayName("thumbnail")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class Thumbnail {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos().create(new VideoCreatePayload().title("[Java-SDK-tests] thumbnail"));
        }

        @Test
        public void uploadThumbnail() throws ApiException {
            File jpgFile = new File(this.getClass().getResource("/assets/cat.jpg").getFile());

            Video video = apiClient.videos().uploadThumbnail(testVideo.getVideoId(), jpgFile);

            assertThat(video.getAssets()).isNotNull();
            assertThat(video.getAssets().getThumbnail()).isNotNull();
        }

        @Test
        public void pickThumbnail() throws ApiException {
            Video video = apiClient.videos().pickThumbnail(testVideo.getVideoId(),
                    new VideoThumbnailPickPayload().timecode("00:00:02"));

            assertThat(video.getAssets()).isNotNull();
            assertThat(video.getAssets().getThumbnail()).isNotNull();
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }

    @Nested
    @DisplayName("video status")
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    class VideoStatus {
        private Video testVideo;

        @BeforeAll
        public void createVideo() throws ApiException {
            this.testVideo = apiClient.videos().create(new VideoCreatePayload().title("[Java-SDK-tests] videoStatus"));
        }

        @Test
        public void getVideoStatus() throws ApiException {
            Videostatus videoStatus = apiClient.videos().getStatus(testVideo.getVideoId());

            assertThat(videoStatus.getIngest()).isNull();
            assertThat(videoStatus.getEncoding()).isNotNull();
            assertThat(videoStatus.getEncoding().getPlayable()).isFalse();
        }

        @AfterAll
        public void deleteVideo() throws ApiException {
            apiClient.videos().delete(testVideo.getVideoId());
        }
    }
}

File: tests/Integration/VideosTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using FluentAssertions;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using ApiVideo.Client;
using ApiVideo.Model;
using ApiVideo.Upload;

namespace VideoApiTests.Integration
{
    public class UploadProgressListener : IUploadProgressListener
    {
        public long totalUploaded = 0;
        public long totalBytes = 0;
        public int chunkCount = 0;
        public List<int> seenChunkNums = new List<int>();

        public void onProgress(long bytesWritten, long totalBytes, int chunkCount, int chunkNum)
        {
            this.totalUploaded = bytesWritten;
            this.totalBytes = totalBytes;
            this.chunkCount = chunkCount;
            this.seenChunkNums.Add(chunkNum);

        }
    }

    [TestClass]
    public class ProgressiveUploadWithTokenTest
    {
        ApiVideoClient apiClient;
        private UploadToken uploadToken;

        [TestInitialize]
        public void init()
        {
            this.apiClient = new ApiVideoClient(System.Environment.GetEnvironmentVariable("API_KEY"));
            this.apiClient.setApplicationName("client-integration-tests", "0");
            this.uploadToken = this.apiClient.UploadTokens().createToken(new TokenCreationPayload());
        }

        [TestMethod]
        public void ProgressiveUpload()
        {
            ApiVideo.Api.VideosApi.UploadWithUploadTokenProgressiveSession uploadWithUploadTokenProgressiveSession = this.apiClient.Videos().createUploadWithUploadTokenProgressiveSession(uploadToken.token);

            var part1 = File.OpenRead("../../../resources/assets/10m.mp4.part.a");
            var part2 = File.OpenRead("../../../resources/assets/10m.mp4.part.b");
            var part3 = File.OpenRead("../../../resources/assets/10m.mp4.part.c");

            uploadWithUploadTokenProgressiveSession.uploadPart(part1);
            uploadWithUploadTokenProgressiveSession.uploadPart(part2);
            Video video = uploadWithUploadTokenProgressiveSession.uploadLastPart(part3);

            part1.Close();
            part2.Close();
            part3.Close();

            this.apiClient.Videos().delete(video.videoid);
        }

        [TestCleanup]
        public void cleanup()
        {
            this.apiClient.UploadTokens().deleteToken(this.uploadToken.token);
        }
    }


    [TestClass]
    public class ProgressiveUploadTest
    {
        ApiVideoClient apiClient;
        Video testVideo;

        [TestInitialize]
        public void init()
        {
            this.apiClient = new ApiVideoClient(System.Environment.GetEnvironmentVariable("API_KEY"));
            this.apiClient.setApplicationName("client-integration-tests", "0");
            this.testVideo = this.apiClient.Videos().create(new VideoCreationPayload() { title = "C# upload stream" });
        }

        [TestMethod]
        public void ProgressiveUpload()
        {
            ApiVideo.Api.VideosApi.UploadProgressiveSession uploadProgressiveSession = this.apiClient.Videos().createUploadProgressiveSession(testVideo.videoid);

            var part1 = File.OpenRead("../../../resources/assets/10m.mp4.part.a");
            var part2 = File.OpenRead("../../../resources/assets/10m.mp4.part.b");
            var part3 = File.OpenRead("../../../resources/assets/10m.mp4.part.c");

            uploadProgressiveSession.uploadPart(part1);
            uploadProgressiveSession.uploadPart(part2);
            Video video = uploadProgressiveSession.uploadLastPart(part3);

            part1.Close();
            part2.Close();
            part3.Close();

        }

        [TestCleanup]
        public void cleanup()
        {
            this.apiClient.Videos().delete(this.testVideo.videoid);
        }
    }


    [TestClass]
    public class VideosUpdateTest
    {
        ApiVideoClient apiClient;
        private Video video;
        private PlayerTheme playerTheme;

        [TestInitialize]
        public void init()
        {
            this.apiClient = new ApiVideoClient(System.Environment.GetEnvironmentVariable("API_KEY"));
            this.apiClient.setApplicationName("client-integration-tests", "0");

            VideoCreationPayload payload = new VideoCreationPayload();
            payload.title = "C# video update";
            this.video = this.apiClient.Videos().create(payload);

            PlayerThemeCreationPayload playerPayload = new PlayerThemeCreationPayload();
            playerPayload.name = "test c#";
            this.playerTheme = this.apiClient.PlayerThemes().create(playerPayload);
        }

        [TestMethod]
        public void Update()
        {
            Video updated;

            VideoUpdatePayload updatePayload = new VideoUpdatePayload();
            updatePayload.playerid = this.playerTheme.playerid;
            updatePayload.title = "player id set";
            updated = this.apiClient.Videos().update(this.video.videoid, updatePayload);
            updated.playerid.Should().Be(this.playerTheme.playerid);

            VideoUpdatePayload updatePayloadUnset = new VideoUpdatePayload();
            updatePayloadUnset.title = "player id not set";
            updated = this.apiClient.Videos().update(this.video.videoid, updatePayloadUnset);
            updated.playerid.Should().Be(this.playerTheme.playerid);

            VideoUpdatePayload updatePayloadNull = new VideoUpdatePayload();
            updatePayloadNull.playerid = null;
            updatePayloadNull.title = "player id null";
            updated = this.apiClient.Videos().update(this.video.videoid, updatePayloadNull);
            updated.playerid.Should().BeNull();
        }

        [TestCleanup]
        public void cleanup()
        {
            this.apiClient.Videos().delete(this.video.videoid);
        }
    }

    [TestClass]
    public class VideosTest
    {
        ApiVideoClient apiClient;
        Video testVideo;

        [TestInitialize]
        public void init()
        {
            this.apiClient = new ApiVideoClient(System.Environment.GetEnvironmentVariable("API_KEY"), ApiVideo.Client.Environment.SANDBOX);
            this.apiClient.setApplicationName("client-integration-tests", "0");

            this.testVideo = apiClient.Videos()
                    .create(new VideoCreationPayload() { title = "[C#-SDK-tests] sdk tests", _public = false});
            Console.WriteLine("Video "+this.testVideo.videoid+" created");
        }
        [TestCleanup]
        public void cleanup()
        {
            this.apiClient.Videos().delete(this.testVideo.videoid);
            Console.WriteLine("Video " + this.testVideo.videoid + " deleted");
        }

        [TestMethod]
        public void UploadVideoByChunk()
        {
            var mp4File = File.OpenRead("../../../resources/assets/10m.mp4");

            long fileSize = mp4File.Length;
            int chunkSize = 1024 * 1024 * 5;

            apiClient.SetUploadChunkSize(chunkSize);
            var upl = new UploadProgressListener();

            apiClient.Videos().upload(testVideo.videoid, mp4File, upl);

            upl.totalBytes.Should().Be(fileSize);
            upl.totalUploaded.Should().Be(fileSize);
            upl.chunkCount.Should().Be((int)Math.Ceiling((float)fileSize / chunkSize));
            upl.seenChunkNums.Should().BeEquivalentTo(1, 2, 3);

            Console.WriteLine(testVideo);
        }

        [TestMethod]
        public void UploadVideoWithoutChunk()
        {
            var mp4File = File.OpenRead("../../../resources/assets/558k.mp4");

            long fileSize = mp4File.Length;
            int chunkSize = (int)fileSize + 1;

            var upl = new UploadProgressListener();

            apiClient.Videos().upload(testVideo.videoid, mp4File, upl);

            upl.totalBytes.Should().Be(fileSize);
            upl.totalUploaded.Should().Be(fileSize);
            upl.chunkCount.Should().Be(1);
            upl.seenChunkNums.Should().BeEquivalentTo(1);

            Console.WriteLine(testVideo);
        }

        [TestMethod]
        public void UpdateAddMetaDataAndTitle()
        {
            Video updated = apiClient.Videos().update(testVideo.videoid,
                    new VideoUpdatePayload()
                    {
                        metadata = new List<Metadata>(){
                            new Metadata(){key="firstKey", value="firstValue" },
                            new Metadata(){key="secondKey", value="secondValue"}
                        }
                    });

            updated.metadata.Should()
                            .BeEquivalentTo(
                                new Metadata(){key="firstKey", value="firstValue" },
                                new Metadata(){key="secondKey", value="secondValue"}
                             );

            Video updated2 = apiClient.Videos().update(testVideo.videoid,
                    new VideoUpdatePayload()
                    {
                        metadata = new List<Metadata>(){
                            new Metadata(){key="firstKey", value="firstValue" },
                            new Metadata(){key="secondKey", value="secondValue"},
                            new Metadata(){key="thirdKey", value="thirdValue"}
                        }
                    });

            updated2.metadata.Should()
                            .BeEquivalentTo(
                                new Metadata() { key = "firstKey", value = "firstValue" },
                                new Metadata() { key = "secondKey", value = "secondValue" },
                                new Metadata() { key = "thirdKey", value = "thirdValue" }
                            );

            var metadata = new Dictionary<string, string>()
            {
                { "firstKey", "firstValue" },
                { "secondKey", "secondValue" },
            };

            Page<Video> page = apiClient.Videos().list().Metadata(metadata).execute();

            page.ItemsTotal.Should().BeGreaterThan(0);

            Video updated3 = apiClient.Videos().update(testVideo.videoid,
                    new VideoUpdatePayload()
                    {
                        metadata = new List<Metadata>()
                    });

            updated3.metadata.Should()
                            .BeEmpty();


        }

        [TestMethod]
        public void Get()
        {
            var video = apiClient.Videos().get(testVideo.videoid);
            video.videoid.Should().Be(this.testVideo.videoid);
            video.title.Should().Be(this.testVideo.title);
        }


        [TestMethod]
        public void uploadThumbnail()
        {
            var jpgFile = File.OpenRead("../../../resources/assets/cat.jpg");

            Video video = apiClient.Videos().uploadThumbnail(testVideo.videoid, jpgFile);

            video.assets.Should().NotBeNull();
            video.assets.thumbnail.Should().NotBeNull();
        }

        [TestMethod]
        public void pickThumbnail()
        {
            var video = apiClient.Videos().pickThumbnail(testVideo.videoid,
                    new VideoThumbnailPickPayload() { timecode = "00:00:02" });

            video.assets.Should().NotBeNull();
            video.assets.thumbnail.Should().NotBeNull();
        }


        // Api side bug I think we have an ingest returned "[]" instead of "{}" so serialization issue
        [TestMethod]
        public void getVideoStatus() 
        {
            VideoStatus videoStatus = apiClient.Videos().getStatus(this.testVideo.videoid);

            videoStatus.ingest.Should().BeNull();
            videoStatus.encoding.Should().NotBeNull();
            videoStatus.encoding.playable.Should().BeFalse();
        }
    }
}
